jms/src/main/scala/akka/stream/alpakka/jms/JmsSourceStage.scala

```diff
@@ -5,80 +5,319 @@
-      private val backpressure = new Semaphore(bufferSize)
+              // By this time, after stopping connection, closing sessions, and closing connections, all async message
+              // submissions to this stage should have been invoked. We invoke markStopped as the last item so it gets
+              // delivered after all JMS messages are delivered. This will allow the stage to complete after all
+              // pending messages are delivered, thus preventing message loss due to premature stage completion.
+              markStopped.invoke(_)
+            } catch {
+              case NonFatal(e) => log.error(e, "Error closing JMS connection {}", jmsConnection)
+            }
+          }
+      }
 
-      private val handleMessage = getAsyncCallback[Message](msg => {
-        require(queue.size <= bufferSize)
-        if (isAvailable(out)) {
-          pushMessage(msg)
-        } else {
-          queue.enqueue(msg)
+  private def abortSessions(ex: Throwable): Unit =
+    if (stopping.compareAndSet(false, true)) {
+      Future {
+        try {
+          jmsConnection.close()
+          log.info("JMS connection {} closed", jmsConnection)
+          markAborted.invoke(ex)
+        } catch {
+          case NonFatal(e) => log.error(e, "Error closing JMS connection {}", jmsConnection)
         }
-      })
+      }
+    }
 
-      override def preStart(): Unit =
-        initSessionAsync(getDispatcher)
+  private[jms] def killSwitch = new KillSwitch {
+    override def shutdown(): Unit = stopSessions()
+    override def abort(ex: Throwable): Unit = abortSessions(ex)
+  }
 
-      private def pushMessage(msg: Message): Unit = {
-        push(out, msg)
-        backpressure.release()
+  override def postStop(): Unit = {
+    queue.clear()
+    stopSessions()
+  }
+}
+
+class NoAckLogic(shape: SourceShape[Message],
+                 out: Outlet[Message],
+                 settings: JmsSourceSettings,
+                 attributes: Attributes)
+    extends SourceStageLogic[Message](shape, out, settings, attributes) {
+
+  private val bufferSize = (settings.bufferSize + 1) * settings.sessionCount
+
+  private val backpressure = new Semaphore(bufferSize)
+
+  private[jms] def createSession(connection: Connection, createDestination: Session => javax.jms.Destination) = {
+    val session =
+      connection.createSession(false, settings.acknowledgeMode.getOrElse(AcknowledgeMode.AutoAcknowledge).mode)
+    new JmsSession(connection, session, createDestination(session))
+  }
+
+  private[jms] def pushMessage(msg: Message): Unit = {
+    push(out, msg)
+    backpressure.release()
+  }
+
+  override private[jms] def onSessionOpened(jmsSession: JmsSession): Unit =
+    jmsSession
+      .createConsumer(settings.selector)
+      .onComplete {
+        case Success(consumer) =>
+          consumer.setMessageListener(new MessageListener {
+            def onMessage(message: Message): Unit = {
+              backpressure.acquire()
```

Comment:
So here we are blocking on a thread that belongs to the JMS provider/client?

Reply:
Yes, this is the same code as before. We do need to block the `onMessage` to pause consumption on that session.